### ISSUE TITLE
systemctl preset and preset-all are inconsistent for instantiated units

### ISSUE THREAD

**Issue by behrmann** [systemd/systemd#15065]
**systemd version the issue has been seen with**
> systemd 244 (244.3-1~bpo10+1)

**Used distribution**
 > Debian buster

**Expected behaviour you didn't see**
 
Having a preset for a unit template of the form `enable mytestunit@foo.service` is inconsistent when using `systemctl preset --preset-mode=full mytestunit@foo.service` and `systemctl preset-all --preset-mode=full` when there is *also* lower priority statement disabling the unit, e.g. `disable *` in `99-default.preset` as in the example from the man page.

The former will enable the unit, whereas the latter will disable it. If one uses the syntax `enable mytestunit@.service foo` from #9901 everything works as expected and both `preset` and `preset-all` will enable the unit.

**Unexpected behaviour you saw**
 
Both `preset` and `preset-all` should enable the unit in the aforementioned case, even when not using the syntax from #9901.

**Steps to reproduce the problem**

Add the following files
```
# /etc/systemd/system/mytestunit@.service
[Unit]
Description=Simple echo unit
[Service]
Type=oneshot
ExecStart=/usr/bin/echo %i
[Install]
WantedBy=multi-user.target
```
```
# /etc/systemd/system-preset/01-local.preset
enable mytestunit@foo.service
```
```
# /etc/systemd/system-preset/99-default.preset
disable *
```

Then one may run
```
root@host:~> systemctl enable mytestunit@foo.service
root@host:~> systemctl is-enabled mytestunit@foo.service 
enabled
root@host:~> systemctl preset-all --preset-mode=full 
Removed /etc/systemd/system/multi-user.target.wants/mytestunit@foo.service.
root@host:~> systemctl is-enabled mytestunit@foo.service 
disabled
root@host:~> systemctl preset --preset-mode=full mytestunit@foo.service
Created symlink /etc/systemd/system/multi-user.target.wants/mytestunit@foo.service → /etc/systemd/system/mytestunit@.service.
root@host:~> systemctl is-enabled mytestunit@foo.service 
enabled
```

Changing the preset file for `mytestunit` to `enable mytestunit@.service foo` everything works as expected
```
root@host:~> systemctl disable mytestunit@foo.service
Removed /etc/systemd/system/multi-user.target.wants/mytestunit@foo.service.
root@host:~> systemctl preset-all --preset-mode=full 
Created symlink /etc/systemd/system/multi-user.target.wants/mytestunit@foo.service → /etc/systemd/system/mytestunit@.service.
root@host:~> systemctl is-enabled mytestunit@foo.service 
enabled
```

**Comment by keszybz** [comment]
Yeah, this seems to be a bug. We need to fix preset-all.

**Comment by behrmann** [comment]
Poking around in the previous discussions on this topic I found this [mailing list post](https://lists.freedesktop.org/archives/systemd-devel/2015-August/033837.html) by @poettering that gives a rationale for the syntax. So, is maybe preset erroneously working when it shouldn't?

Also, having looked at the code a bit, is there a deeper reason why `read_presets` only uses `split_pattern_into_name_and_instances` for enable and not for disable?

**Comment by boucman** [comment]
yeah... it's not very clear, but enable and disable have completely different logic

*  Enable will read the [Install] section and do the actions described in there
* Disable will suppress all symlinks in /etc that point to the unit that is being disabled.

Most of the time, the two actions are symetrical and you don't notice the difference, but if you edit a service, this logic guarantees that the unit will be properly disabled, even when the Install section has changed and we don't really know how the unit was enabled in the first place...

Unfortunately this logic kinda breaks with templates. Disabling a template will disable all instances of the template, where you would expect it to only disable a DefaultInstance. But there is no safe way to know what the DefaultInstance is/was...

**Comment by behrmann** [comment]
Thanks for the clarification, @boucman, I didn't know that about enable vs disable! But I guess my question was a bit imprecise. 

If I see it correctly, `split_pattern_into_name_and_instances` is only used in the enable part of the logic, which I understand as the `mytemplate@.service foo bar baz` syntax will only work in the case `enable mytemplate@.service foo bar baz` but not for `disable mytemplate@.service foo bar baz`. 

But instances are symlinks under /etc as well, are they not? So it would not run counter the above logic of suppressing the appropriate symlinks if one were to expand the usage of `split_pattern_into_name_and_instances` into the disable branch in [`read_presets`](https://github.com/systemd/systemd/blob/master/src/shared/install.c#L2917)?

Also, this is a bit orthogonal to the initial issue. I guess a place to fix that would be `split_pattern_name_and_instances` to extend to get a single instance from between `@` and `.`, even though that would run against the initial rationale for the introduction of the list syntax, it would keep the compatibility to the whole thing already working for `preset`. Does that sound sensible?

**Comment by clarfonthey** [comment]
FWIW, this bug makes instanced units particularly unusable for presets since the instances are overridden, not added, in the case of presets.

Essentially, the below two lines cancel each other out:

```
enable thing@.service one
enable thing@.service two
```

whereas the below two lines don't:

```
enable thing-one.service
enable thing-two.service
```

Essentially, users are encouraged to copy-paste units so they all have distinct names instead of using instances, since presets don't really honour instances at all and only take the last line read as the canonical set of instances. Alternatively, instead of using presets, you have to just create your own unit drop-in files to add `Wants=…` lines for all the necessary instances, which kind of defeats the purpose of presets.

I tried looking at the code to figure out where exactly the current behaviour is coming from but gave up after finding that the preset code is kind of a giant spaghetti mess.

**Comment by behrmann** [comment]
What you're describing is not this issue exactly, because
```
enable thing@.service one
enable thing@.service two
```
should be
```
enable thing@.service one two
```

**Comment by clarfonthey** [comment]
My point is that presets are explicitly allowed to be spread between multiple files and forcing instances to specifically be configured in exactly one line across all of those files is kind of antithetical to how presets are designed.
